docs/design/three_datahall.md

```diff
@@ -0,0 +1,204 @@
+For on-premise clusters this could be a rack, a datahall or a data center.
+The current implementation of the operator only supports either single FoundationDB clusters with the redundancy modes `single`, `double`, `triple` or HA clusters span across multiple Kubernetes clusters.
+FoundationDB offers the redundancy mode [three_data_hall](https://apple.github.io/foundationdb/configuration.html#single-datacenter-modes) or [three_datacenter](https://apple.github.io/foundationdb/configuration.html#datacenter-aware-mode) that fits well in these enviornments without the overhead of the HA solution.
+The current requirement for `three_data_hall` and `three_datacenter` is to have 3 AZs and at least two different failure zones per `data_hall`/`datacenter`.
+
+## General Design Goals
+
+The goal of this design is to support the redundancy modes `three_data_hall` and `three_datacenter` in different deployment models:
+
+* Single Kubernetes cluster across multiple AZs.
+* One Kubernetes cluster per AZ.
+
+The implementation should be flexible enough to give the user a choice of the used AZ.
+
+## Current Implementation
+
+The current version of the operator doesn't support this redundancy mode.
+The locality is currently a mix of some constants, a configurable zone and two settings at the cluster level [locality configuration](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/master/controllers/cluster_controller.go#L631-L642).
+
+## Proposed Design
+
+The design is split into two main parts of the deployment: configure the locality and how to deploy the Pods across the cluster.
+
+### Configure locality
+
+The `three_data_hall` deployment requires the `locality_data_hall` locality to be set and the `three_datacenter` requires the `locality_dcid` to be set.
+There must be at least and at most 3 different values for these localities.
+To full fill the requirements there must be at least two different `locality_zone`s per `data_hall`/`datacenter`.
+The `locality_zone` can be configured with [FoundationDBClusterFaultDomain](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/master/docs/cluster_spec.md#foundationdbclusterfaultdomain).
+In addition to that the user can configure additional variables that the sidecar will use for substitution in the [FoundationDBClusterSpec](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/master/docs/cluster_spec.md#foundationdbclusterspec) with the `sidecarVariables` list.
+The current implementation would allow to set `dataHall: $AZ` and in `sidecarVariables` we would list `AZ` to define the `locality_data_hall` based on an environment variable that will be replaced by the sidecar with the actual value.
+Instead of having these different mechanisms spread across different settings I propose to have a new `localities` setting in the `FoundationDBClusterSpec`.
+The key will always be prefixed with `locality_`:
+
+```yaml
+localities:
+- key: "data_hall"
+  value: ""
```

Comment:
If this is fully replacing the FaultDomain structure, do we need a special mechanism to mimic the behavior of the `foundationdb.org/none` fault domain strategy, or is that something we can reproduce with `valueFromEnv`?

Reply:
We also have logic to set environment variables for fault domain information based on fields on the pod, which is how we inject the node name today. Is that kind of behavior going to be handled through `valueFromNode` instead?

Reply:
This would be something we could mimic with `valueFromEnv` and then use use the Pod name. If those fields are "on the Pod" I think that would be a case for `valueFromEnv` the idea of `valueFromNode` is to read labels from the node that "hosts" the specific Pod.